pkg/service/controller/servingcert/secret_updating_controller.go

```diff
@@ -0,0 +1,283 @@
+		secret = old.(*kapi.Secret)
+		if len(secret.Annotations[ServiceNameAnnotation]) == 0 {
+			return
+		}
+	}
+
+	glog.V(4).Infof("updating %s", secret.Name)
+	sc.enqueueSecret(secret)
+}
+
+func (sc *ServiceServingCertUpdateController) runWorker() {
+	for sc.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (sc *ServiceServingCertUpdateController) processNextWorkItem() bool {
+	key, quit := sc.queue.Get()
+	if quit {
+		return false
+	}
+	defer sc.queue.Done(key)
+
+	err := sc.syncHandler(key.(string))
+	if err == nil {
+		sc.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+	sc.queue.AddRateLimited(key)
+
+	return true
+}
+
+// syncSecret will sync the service with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (sc *ServiceServingCertUpdateController) syncSecret(key string) error {
+	obj, exists, err := sc.secretCache.GetByKey(key)
+	if err != nil {
+		glog.V(4).Infof("Unable to retrieve service %v from store: %v", key, err)
+		return err
+	}
+	if !exists {
+		glog.V(4).Infof("Secret has been deleted %v", key)
+		return nil
+	}
+
+	if !sc.requiresRegeneration(obj.(*kapi.Secret)) {
+		return nil
+	}
+
+	// make a copy to avoid mutating cache state
+	t, err := kapi.Scheme.DeepCopy(obj)
+	if err != nil {
+		return err
+	}
+	secret := t.(*kapi.Secret)
+
+	dnsName := secret.Annotations[ServiceNameAnnotation] + "." + secret.Namespace + ".svc"
+	fqDNSName := dnsName + "." + sc.dnsSuffix
+	servingCert, err := sc.ca.MakeServerCert(sets.NewString(dnsName, fqDNSName))
+	if err != nil {
+		return err
+	}
+	secret.Annotations[ServingCertExpiryAnnotation] = servingCert.Certs[0].NotAfter.Format(time.RFC3339)
+	secret.Data[kapi.TLSCertKey], secret.Data[kapi.TLSPrivateKeyKey], err = servingCert.GetPEMBytes()
+	if err != nil {
+		return err
+	}
+
+	_, err = sc.secretClient.Secrets(secret.Namespace).Update(secret)
+	return err
+}
+
+func (sc *ServiceServingCertUpdateController) requiresRegeneration(secret *kapi.Secret) bool {
+	serviceName := secret.Annotations[ServiceNameAnnotation]
+	if len(serviceName) == 0 {
+		return false
+	}
+
+	serviceObj, exists, err := sc.serviceCache.GetByKey(secret.Namespace + "/" + serviceName)
+	if err != nil {
+		return false
+	}
+	if !exists {
+		return false
+	}
+
+	service := serviceObj.(*kapi.Service)
+	if secret.Annotations[ServiceUIDAnnotation] != string(service.UID) {
+		return false
+	}
+
+	// if we don't have the annotation for expiry, just go ahead and regenerate.  It's easier than writing a
+	// secondary logic flow that creates the expiry dates
+	expiryString, ok := secret.Annotations[ServingCertExpiryAnnotation]
+	if !ok {
+		return true
+	}
+	expiry, err := time.Parse(time.RFC3339, expiryString)
+	if err != nil {
+		return true
+	}
+
+	if expiry.Add(sc.minTimeLeftForCert).After(time.Now()) {
```

Comment:
> Is this a correct calculation?
> 
> As far I understand certificate requires regeneration if now > (expiration date - 1h) IOW we'll update certificate if it expires in next hour. But here we're adding 1 hour instead of subtracting it. What I'm missing here?
> 
> Also: why 1 hour if we're checking each 20 minutes?

heh, backwards.  I'll add @soltysh unit test and catch it. :)

The 1 hour is a place holder for something more reasonable driven off a "real" expiry time.  I suspect we'll have an upper bound too to eliminate all long lived tokens.